php/class-customize-concurrency.php

```diff
@@ -14,4 +14,284 @@
+				'post_date' => current_time( 'mysql', 0 ),
+				'post_date_gmt' => current_time( 'mysql', 1 ),
+				'post_author' => get_current_user_id(),
+			);
+
+			if ( $is_update ) {
+				$post_data['ID'] = $saved_settings[ $setting_id ]['post_id'];
+				$r[] = wp_update_post( wp_slash( $post_data ), true );
+			} else {
+				$r[] = wp_insert_post( wp_slash( $post_data ), true );
+			}
+		}
+
+		$this->restore_kses();
+		// Todo: check $r for errors
+	}
+
+	public function customize_save_response( $response ) {
+		$response['concurrency_session_timestamp'] = strtotime( current_time( 'mysql', true ) );
+		return $response;
+	}
+
+	/**
+	 * Get saved settings from default post storage or from snapshot storage.
+	 *
+	 * Todo: read from snapshots when applicable.
+	 *
+	 * @return array
+	 */
+	function get_saved_settings() {
+		$saved_settings = array();
+
+		add_filter( 'sanitize_title', array( $this, 'sanitize_title_for_query' ), 10, 3 );
+		$saved_setting_posts = new \WP_Query(array(
+			'post_name__in' => array_keys( $this->customize_manager->unsanitized_post_values() ),
+			'post_type' => self::POST_TYPE,
+			'post_status' => 'publish',
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'posts_per_page' => -1,
+			'ignore_sticky_posts' => true,
+		));
+		remove_filter( 'sanitize_title', array( $this, 'sanitize_title_for_query' ) );
+
+		foreach ( $saved_setting_posts->posts as $setting_post ) {
+
+			$saved_settings[ $setting_post->post_name ] = array(
+				'post_id' => $setting_post->ID,
+				'value' => json_decode( $setting_post->post_content_filtered, true ),
+				'timestamp' => strtotime( $setting_post->post_modified_gmt ),
+				'author' => $setting_post->post_author,
+			);
+		}
+
+		return $saved_settings;
+	}
+
+	/**
+	 * Suspend kses which runs on content_save_pre and can corrupt JSON in post_content.
+	 *
+	 * @see \sanitize_post()
+	 */
+	function suspend_kses() {
+		if ( false !== has_filter( 'content_save_pre', 'wp_filter_post_kses' ) ) {
+			$this->kses_suspended = true;
+			kses_remove_filters();
+		}
+	}
+
+	/**
+	 * Restore kses which runs on content_save_pre and can corrupt JSON in post_content.
+	 *
+	 * @see \sanitize_post()
+	 */
+	function restore_kses() {
+		if ( $this->kses_suspended ) {
+			kses_init_filters();
+			$this->kses_suspended = false;
+		}
+	}
+
+	/**
+	 * Keep square brackets from being removed when the title is sanitized for the query in $this->get_saved_settings()
+	 *
+	 * @see get_saved_settings()
+	 * @filter sanitize_title
+	 *
+	 * @param $title     string Sanitized title without the needed brackets
+	 * @param $raw_title string Original title to revert to if the only difference is square brackets
+	 * @param $context   string
+	 * @return string
+	 */
+	function sanitize_title_for_query( $title, $raw_title, $context ) {
+		if ( 'query' === $context && preg_replace( '/[\[\]]/', '', $raw_title ) === $title ) {
```

Comment:
It's not really about the performance gain. It's more about just being sure that the original title being used. There could be another `sanitize_title` filter that forces `-` instead of `_`, for example.  So also, actually, you should increase the priority to be like 1000 maybe.

Something else needed here is `esc_sql()`, unfortunately. See another implementation of this filter here: https://github.com/xwp/wp-customize-widgets-plus/blob/d33bdac61934c303c328aa8c0af8cc212be490be/php/class-widget-posts.php#L1153-L1170

The reason being is that the return value for `sanitize_title_for_query()` will get supplied straight into a SQL string (somewhat baffling), so we have to make sure that it is escaped for SQL: https://github.com/xwp/wordpress-develop/blob/1ca2426e779b8e1f25d80e2cfcf1e3c05ddb41f0/src/wp-includes/query.php#L2716-L2718